include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -228,71 +228,411 @@ struct __merge_sort_leaf_submitter<__internal::__optional_kernel_name<_LeafSortN
+        //   |              |              |              |              |              |              |              |         |    |              |              |              |              |
+        // bd00           bd01           bd02           bd10           bd11           bd12           bd20           bd21        |  bd22           bd30           bd31           bd32           bd40                              < Base diagonals
+        //                  ^              ^              ^              ^              ^              ^              ^         |    ^              ^              ^              ^              ^
+        //  ---             0              1              2              3              4              5              6         |    7              8              9              10             11                              < Indexes in the base diagonal's SP storage
+        //   0    1    2    3    4    5    6    7    8    9    10   11   12   13   14  15   16    17   18   19   20   20   21   |    23   24   25   26   27   28   29   30   31   32   33   34   35    36                        < Linear IDs: __linear_id_in_steps_range
+        //   ^                                                                                                         |         |    |
+        //   |                                                                                                       __sp_left   |  __sp_right
+        //   |                                                                                                                   |
+        //   |                                                                                                       __linear_id_in_steps_range
+        //  We don't save the first diagonal into base diagonal's SP storage !!!
+
+        std::size_t __diagonal_idx = __linear_id_in_steps_range / __nd_range_params.steps_between_two_base_diags;
+
+        assert(__diagonal_idx < __nd_range_params.base_diag_count);
+
+        const _merge_split_point_t __sp_left =
+            __diagonal_idx > 0 ? __base_diagonals_sp_global_ptr[__diagonal_idx - 1] : _merge_split_point_t{0, 0};
+        const _merge_split_point_t __sp_right = __base_diagonals_sp_global_ptr[__diagonal_idx];
+
+        const bool __is_base_diagonal =
+            __linear_id_in_steps_range % __nd_range_params.steps_between_two_base_diags == 0;
+
+        return __sp_right.first + __sp_right.second > 0
+                   ? (!__is_base_diagonal
+                          ? __find_start_point(__views.rng1, __sp_left.first, __sp_right.first, __views.rng2,
+                                               __sp_left.second, __sp_right.second, __data_area.i_elem_local, __comp)
+                          : __sp_left)
+                   : __find_start_point_w(__data_area, __views, __comp);
```

Comment:
I don't know about anyone else, but this is so much harder to read for me than something like this:

```suggestion
    if (__sp_right.first + __sp_right.second > 0)
    {
        if (!__is_base_diagonal)
        {
            return __find_start_point(__views.rng1, __sp_left.first, __sp_right.first,
                                                    __views.rng2, __sp_left.second, __sp_right.second,
                                                    __data_area.i_elem_local,
                                                    __comp);
        }
        else
            return __sp_left;
    }
    else 
        return __find_start_point_w(__data_area, __views, __comp);      
```
Others don't respond in agreement, then I'm fine to leave it as is.

Reply:
Fixed